### DISCUSSION TITLE
Can a route influence a middleware (for access control)?

### DISCUSSION THREAD

**Discussion by jessemyers-lettuce** [fastapi/fastapi#9899] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, HTTPException, FastAPI, Request, status

app = FastAPI()
app.state.allowed = False


@app.middleware("http")
async def foo(request: Request, call_next):

    if not app.state.allowed:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Forbidden")

    response = await call_next(request)
    return response

def bar() -> None:
    # I want this to be called early enough that the `foo` can see the result
    app.state.allowed = True


@app.get("/", dependencies=[Depends(bar)])
async def root(request: Request):
    return {"message": "Hello World"}
```

### Description

I'd like to define access control logic where the **default** behavior is to **reject** a request, but individual routes can allow access based on <some rule>.

I am willing to decorate/annotate every route with some metadata (or dependency) that requires a specific behavior, but I don't want the **default**behavior -- if I forget this metadata/dependency - to be to **allow** an operation.

My ideal solution is:

 - A middleware checks a property of `Request` to determine whether the request is allowed.
 - A route can update this property to allow the operation (or allow it conditionally based on some function)
 - If the route does not update this property or updates the property with a function that is not satisfied, the middleware rejects the request (e.g. with a `403 Forbidden`)
 - Otherwise, the route is allowed to continue.

But... I can't see any way to have a route influence a middleware (since they are resolved earlier) and I can't see anyway to allow a middleware to know what route would be called (because middleware can be stacked).

This leaves me without a good solution. I can build a mediocre solution by decorating/annotating every route, but I can't make the **default** behavior to reject access.

### Operating System

macOS

### FastAPI Version

0.100.0

### Python Version

Python 3.10.11

### Additional Context

I suspect that what I want is not possible, but I'm looking for confirmation before I give up.

**Comment by jgould22** [comment]
What is the intention with having a route that returns a 403 by default, why have the route in this case ?

Anyway

You can probably use 2 routers, one with a global dependency that rejects. 

When you want that behaviour to change you can attach the route to the router without the global dependency that rejects.

**Reply by jessemyers-lettuce** [reply]
@jgould22 The intent is to implement access control using a "default deny" policy. 

In many systems, access control rules for different roles are contingent on the user role *and* route-specific factors. For example, admin users might be able to query for all users in a system, manager users might only be able to query for users in their organization, and regular users might only be able to query for themselves.

It is, of course, possible to write these rules as part of route logic; the endpoint that enable user search can inspect user rules and query parameters and make an appropriate decision. However, it's common to forget/neglect this logic and write the endpoint initially with overly=permissive logic. I'd like to find a way to both avoid this problem and make these rules very explicit.

One possible implementation would be a classic ACL system: literally a list of access rules. My hope was to write this system such that a middleware or router dependency had a default list of rules (e.g. allow if admin, otherwise deny) and then allow specific routes to supply their own rules (e.g. allow this route for all users; allow this route if a manager and the organization is correctly defined).

The problem is that I don't think any behavior that can be made to apply to all routes *unconditionally* gets visibility into behavior that routes can supply individually. I could move the enforcement to the route dependencies, but that defeats the purpose of the "default deny" policy; if a route omits this logic, the entire enforcement mechanism goes away.

**Comment by iudeen** [comment]
You can have a set of access/control rules defined externally (in a db like redis). Your middleware or dependency can then check from that datastore/cache.

**Reply by jessemyers-lettuce** [reply]
@iudeen I understand that part; what I'm missing is a way for the middleware to know which route will be called at the end of the middleware chain so it can access route-specific details.

**Reply by iudeen** [reply]
You can get it from ASGI [Scope](https://asgi.readthedocs.io/en/latest/specs/main.html#connection-scope:~:text=Closely%20related%2C%20the%20first%20argument%20passed%20to%20an%20application%20callable%20is%20a%20scope%20dictionary%20with%20all%20the%20information%20describing%20that%20specific%20connection.)

I wrote a rough code (not the cleanest) for your reference. 

```python
import typing

from fastapi import FastAPI

Scope = typing.MutableMapping[str, typing.Any]
Message = typing.MutableMapping[str, typing.Any]

Receive = typing.Callable[[], typing.Awaitable[Message]]
Send = typing.Callable[[Message], typing.Awaitable[None]]

ASGIApp = typing.Callable[[Scope, Receive, Send], typing.Awaitable[None]]


class MyMiddlware:
    def __init__(
            self,
            app,
    ):
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope["type"] != "http":
            """
            this middleware only handles http scope.
            Ignore other ASGI compatible scopes (lifespan and websockets) here.
            """
            return await self.app(scope, receive, send)

        print(scope.get("path"))
        await self.app(scope, receive, send)


app = FastAPI()

app.add_middleware(MyMiddlware)


@app.get("/")
def get():
    return "hello"


@app.get("/hello")
def get2():
    return "hello"
```

**Reply by jessemyers-lettuce** [reply]
@iudeen That's certainly a help. I was really hoping to get a reference to the endpoint function (`get` or `get2`)  and not just the path, but maybe there's a way to interrogate the router.

**Reply by jessemyers-lettuce** [reply]
It looks like I can:

```py
for route in self.app.routes:
    match, child_scope = route.matches(scope)
    if match == Match.FULL:
        return child_scope["endpoint"]

return None
```

There appears to be support in `starlette` routing for partial matches and I haven't figured out how that works yet; in my tests so far, I only seem to care about full matches.

**Reply by jessemyers-lettuce** [reply]
Looking at the router source, partial appears to mean "the path matched but the method did not", which isn't relevant in my case.

**Comment by jessemyers-lettuce** [comment]
Thanks for the help, I have enough to move forward.
